Add --list and --cleanup options to snapshot command, for #1163

cmuench opened the pull request:
## The Problem/Issue/Bug:
I want be able to list and delete existing db-snapshots. A command or option is missing.

## How this PR Solves The Problem:
Add a `--list` option to snapshot command.
Add a `--cleanup` option to snapshot command.

## Release/Deployment notes:
Code should not affect any existing project.

Related snapshot and restore-snapshot issues are at https://github.com/drud/ddev/issues?q=is%3Aissue+is%3Aopen+snapshot

rfay commented:
Cool - if you're going to be this ambitious, would you like to take on all of https://github.com/drud/ddev/issues/1886 ?

cmuench commented:
@rfay you mean to add the "restore-snapshot --latest" option? Yes, maybe

rfay commented:
Sure. Maybe a snapshot cleanup? I think there are some other suggestions in that issue?

cmuench commented:
@rfay ok. Will try to add the `--cleanup` option as next PR.

cmuench commented:
@rfay I added the `--cleanup` option to the snapshot command.
The cleanup supports the following:

- Cleanup a single snapshot: `ddev snapshot  --cleanup --name <snapshotname> [projectname]`
- Cleanup all existing snapshots of a project: `ddev snapshot --cleanup [projectname]`
- Cleanup all existing snapshots of all projects: `ddev snapshot --cleanup --all`

I also added a test for the snapshot delete method of the ddev app.

rfay reviewed (changes requested):
Wow, this is awesome! You really are ambitious. I manually tested and loved the results. 

* Some trivial suggestions that you can just commit (as a batch is best).
* I think it needs a test in the cmd section to verify that flags work correctly, see cmd/ddev/cmd/start_test.go for a strategy. 
* Needs a rebase as well.

If you'd be willing to do the test I think we'd be in great shape. Awesome work on this. Love the results. Hoping you'll work on restore-snapshot next!

rfay commented on cmd/ddev/cmd/snapshot.go:
```diff
@@ -26,17 +32,59 @@ ddev snapshot --all`,
+			case snapshotList:
+				listAppSnapshots(app)
+			case snapshotCleanup:
+				deleteAppSnapshot(app)
```
I wondered if this should have a util.Confirm() before doing it. Should people be asked before deleting all the snapshots?

cmuench replied:
Any deletion requires to add an explicit option. So it cannot happen accidentally.
Maybe we should add the confirmation in combination with the  "--all" option?
If we add this, it would be useful to have a `--force` option as well.

rfay replied:
There's a global --yes/-y option. We'd just have to check that. 

You say 
> Any deletion requires to add an explicit option

but I'd say that a util.Confirm here would be good, overridden with the default --yes/-y

rfay replied:
Weird, I'm pretty sure it's supposed to be a global flag, but the best example for you (in delete.go) doesn't do it globally at all.

cmuench replied:
I added a confirmation option. The delete command additionally forbid to combine `--all` and `--yes`. I did not port that logic, because it's useful to cleanup all snapshots. What do you think?

rfay commented:
Related snapshot and restore-snapshot issues: https://github.com/drud/ddev/issues?q=is%3Aissue+is%3Aopen+snapshot

rfay commented:
I went ahead and rebased this to deal with some test problems, so make sure you update to your remote before doing anything...

cmuench commented on pkg/ddevapp/ddevapp_test.go:
```diff
@@ -1367,6 +1367,47 @@ func TestDdevFullSiteSetup(t *testing.T) {
 }
 
+// TestDdevSnapshotDelete tests creating a snapshot and delete it. This runs with Mariadb 10.2
+func TestDdevSnapshotCleanup(t *testing.T) {
```
I am a test fan.

cmuench commented:
@rfay Docker for windows NFS build is broken. I cannot see the build log.

rfay commented:
I restarted it; looked to me like a "normal" windows docker fail.  If you get me the email associated with your github username I'll invite you to buildkite.

cmuench commented:
> I restarted it; looked to me like a "normal" windows docker fail. If you get me the email associated with your github username I'll invite you to buildkite.

christian@muench-worms.de

rfay commented:
Invite is in your email. thanks!

The Windows builds are relatively notorious. Docker breaks sometimes, etc.

rfay reviewed (approved):
This is lovely, and works fine. The one unanticipated thing is that a command against a project with multiple snapshots requires a confirmation for each snapshot. Is that what you'd expected? Of course, one could still use -y to override it. But I would have expected a single confirmation to delete all of them.

So for example, d9composer has several snapshots. Here's what happens:

```
$ ddev snapshot d9composer --cleanup
OK to delete the snapshot d9composer_20201127063839 of d9composer
. [Y/n] (yes):
Deleted database snapshot d9composer_20201127063839 in /Users/rfay/workspace/d9composer/.ddev/db_snapshots/d9composer_20201127063839
OK to delete the snapshot d9composer_20201127064005 of d9composer
. [Y/n] (yes):
Deleted database snapshot d9composer_20201127064005 in /Users/rfay/workspace/d9composer/.ddev/db_snapshots/d9composer_20201127064005
OK to delete the snapshot d9composer_20201127065224 of d9composer
. [Y/n] (yes):
Deleted database snapshot d9composer_20201127065224 in /Users/rfay/workspace/d9composer/.ddev/db_snapshots/d9composer_20201127065224
```

If you're fine with that (and it's not unreasonable behavior) then this is ready to go.  Thanks for your outstanding work on this!

cmuench commented:
@rfay I thought it's more safe. I tested a bit with it and agree with you. It's more handy to have one confirmation per project.

cmuench commented:
@rfay ready to merge.

rfay reviewed (changes requested):
Sorry, just one last nit, with confirmation formatting, that line break shouldn't be there before the confirmation.
```
rfay@rfay-mbp-2017:~/tmp/ddev/.gotmp/bin/darwin_amd64$ ./ddev snapshot --cleanup -a
OK to delete all snapshots of d9
. [Y/n] (yes): ^C
rfay@rfay-mbp-2017:~/tmp/ddev/.gotmp/bin/darwin_amd64$ ./ddev snapshot --cleanup d9
OK to delete all snapshots of d9
. [Y/n] (yes):
```

cmuench commented:
> Sorry, just one last nit, with confirmation formatting, that line break shouldn't be there before the confirmation.
> 
> ```
> rfay@rfay-mbp-2017:~/tmp/ddev/.gotmp/bin/darwin_amd64$ ./ddev snapshot --cleanup -a
> OK to delete all snapshots of d9
> . [Y/n] (yes): ^C
> rfay@rfay-mbp-2017:~/tmp/ddev/.gotmp/bin/darwin_amd64$ ./ddev snapshot --cleanup d9
> OK to delete all snapshots of d9
> . [Y/n] (yes):
> ```

Removed

rfay reviewed (approved):
Fantastic, thanks! I'll pull it when the tests are done. It will go into v1.7.0-alpha2
